File: website/algorithms/kadane-s-algorithm/Kadanes-Algorithm.js
/**
 * Kadane's Algorithm Implementation in JavaScript
 * 
 * Kadane's Algorithm is used to find the maximum sum of a contiguous subarray 
 * within a one-dimensional array of numbers in O(n) time complexity.
 * 
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * Basic Kadane's Algorithm to find maximum sum of contiguous subarray
 * @param {number[]} arr - Input array
 * @returns {number} Maximum sum of contiguous subarray
 */
function kadanesAlgorithm(arr) {
    if (arr.length === 0) return 0;

    let maxSoFar = Number.NEGATIVE_INFINITY;
    let maxEndingHere = 0;

    for (let i = 0; i < arr.length; i++) {
        maxEndingHere = maxEndingHere + arr[i];

        if (maxSoFar < maxEndingHere) {
            maxSoFar = maxEndingHere;
        }

        if (maxEndingHere < 0) {
            maxEndingHere = 0;
        }
    }

    return maxSoFar;
}

/**
 * Kadane's Algorithm that also returns the subarray indices
 * @param {number[]} arr - Input array
 * @returns {Object} Object containing maxSum, startIndex, and endIndex
 */
function kadanesWithIndices(arr) {
    if (arr.length === 0) return { maxSum: 0, startIndex: -1, endIndex: -1 };

    let maxSoFar = Number.NEGATIVE_INFINITY;
    let maxEndingHere = 0;
    let start = 0;
    let end = 0;
    let s = 0; // temporary start index

    for (let i = 0; i < arr.length; i++) {
        maxEndingHere += arr[i];

        if (maxSoFar < maxEndingHere) {
            maxSoFar = maxEndingHere;
            start = s;
            end = i;
        }

        if (maxEndingHere < 0) {
            maxEndingHere = 0;
            s = i + 1;
        }
    }

    return {
        maxSum: maxSoFar,
        startIndex: start,
        endIndex: end,
        subarray: arr.slice(start, end + 1)
    };
}

/**
 * Alternative implementation using reduce
 * @param {number[]} arr - Input array
 * @returns {number} Maximum sum of contiguous subarray
 */
function kadanesReduce(arr) {
    if (arr.length === 0) return 0;

    let maxSoFar = Number.NEGATIVE_INFINITY;

    arr.reduce((maxEndingHere, current) => {
        maxEndingHere = Math.max(current, maxEndingHere + current);
        maxSoFar = Math.max(maxSoFar, maxEndingHere);
        return maxEndingHere;
    }, 0);

    return maxSoFar;
}

/**
 * Kadane's Algorithm for circular array (bonus implementation)
 * @param {number[]} arr - Input array
 * @returns {number} Maximum sum of contiguous subarray in circular array
 */
function kadanesCircular(arr) {
    if (arr.length === 0) return 0;

    // Case 1: Maximum subarray is non-circular
    const maxKadane = kadanesAlgorithm(arr);

    // Case 2: Maximum subarray is circular
    // Find minimum subarray sum and subtract from total
    const totalSum = arr.reduce((sum, num) => sum + num, 0);

    // Invert array signs to find minimum subarray using Kadane's
    const invertedArr = arr.map(num => -num);
    const maxInverted = kadanesAlgorithm(invertedArr);
    const minSubarray = -maxInverted;

    const maxCircular = totalSum - minSubarray;

    // If all elements are negative, return maxKadane
    return maxCircular === 0 ? maxKadane : Math.max(maxKadane, maxCircular);
}

// Example usage and testing
function main() {
    console.log("Kadane's Algorithm Examples:\n");

    // Test case 1: Mixed positive and negative numbers
    const arr1 = [-2, -3, 4, -1, -2, 1, 5, -3];
    console.log("Array 1:", arr1);
    console.log("Maximum contiguous sum:", kadanesAlgorithm(arr1));
    console.log("With indices:", kadanesWithIndices(arr1));
    console.log();

    // Test case 2: All negative numbers
    const arr2 = [-2, -3, -1, -5];
    console.log("Array 2:", arr2);
    console.log("Maximum contiguous sum:", kadanesAlgorithm(arr2));
    console.log();

    // Test case 3: All positive numbers
    const arr3 = [1, 2, 3, 4, 5];
    console.log("Array 3:", arr3);
    console.log("Maximum contiguous sum:", kadanesAlgorithm(arr3));
    console.log();

    // Test case 4: Circular array
    const arr4 = [5, -3, 5];
    console.log("Array 4 (circular):", arr4);
    console.log("Maximum contiguous sum (linear):", kadanesAlgorithm(arr4));
    console.log("Maximum contiguous sum (circular):", kadanesCircular(arr4));
    console.log();

    // Performance comparison
    console.log("Reduce implementation result:", kadanesReduce(arr1));
}

// Run the example
main();

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        kadanesAlgorithm,
        kadanesWithIndices,
        kadanesReduce,
        kadanesCircular
    };
}

File: website/algorithms/kadane-s-algorithm/Kadanes-Algorithm.py
"""
Kadane's Algorithm Implementation in Python

Kadane's Algorithm is used to find the maximum sum of a contiguous subarray 
within a one-dimensional array of numbers in O(n) time complexity.

Time Complexity: O(n)
Space Complexity: O(1)
"""

def kadanes_algorithm(arr):
    """
    Basic Kadane's Algorithm to find maximum sum of contiguous subarray
    
    Args:
        arr (list): Input array of integers
        
    Returns:
        int: Maximum sum of contiguous subarray
    """
    if not arr:
        return 0
    
    max_so_far = float('-inf')
    max_ending_here = 0
    
    for num in arr:
        max_ending_here = max_ending_here + num
        
        if max_so_far < max_ending_here:
            max_so_far = max_ending_here
        
        if max_ending_here < 0:
            max_ending_here = 0
    
    return max_so_far

def kadanes_with_indices(arr):
    """
    Kadane's Algorithm that also returns the subarray indices
    
    Args:
        arr (list): Input array of integers
        
    Returns:
        dict: Dictionary containing max_sum, start_index, end_index, and subarray
    """
    if not arr:
        return {'max_sum': 0, 'start_index': -1, 'end_index': -1, 'subarray': []}
    
    max_so_far = float('-inf')
    max_ending_here = 0
    start = 0
    end = 0
    s = 0  # temporary start index
    
    for i, num in enumerate(arr):
        max_ending_here += num
        
        if max_so_far < max_ending_here:
            max_so_far = max_ending_here
            start = s
            end = i
        
        if max_ending_here < 0:
            max_ending_here = 0
            s = i + 1
    
    return {
        'max_sum': max_so_far,
        'start_index': start,
        'end_index': end,
        'subarray': arr[start:end + 1]
    }

def kadanes_pythonic(arr):
    """
    Pythonic implementation using built-in functions
    
    Args:
        arr (list): Input array of integers
        
    Returns:
        int: Maximum sum of contiguous subarray
    """
    if not arr:
        return 0
    
    max_so_far = float('-inf')
    max_ending_here = 0
    
    for num in arr:
        max_ending_here = max(num, max_ending_here + num)
        max_so_far = max(max_so_far, max_ending_here)
    
    return max_so_far

def kadanes_circular(arr):
    """
    Kadane's Algorithm for circular array (bonus implementation)
    
    Args:
        arr (list): Input array of integers
        
    Returns:
        int: Maximum sum of contiguous subarray in circular array
    """
    if not arr:
        return 0
    
    # Case 1: Maximum subarray is non-circular
    max_kadane = kadanes_algorithm(arr)
    
    # Case 2: Maximum subarray is circular
    # Find minimum subarray sum and subtract from total
    total_sum = sum(arr)
    
    # Invert array signs to find minimum subarray using Kadane's
    inverted_arr = [-num for num in arr]
    max_inverted = kadanes_algorithm(inverted_arr)
    min_subarray = -max_inverted
    
    max_circular = total_sum - min_subarray
    
    # If all elements are negative, return max_kadane
    return max_kadane if max_circular == 0 else max(max_kadane, max_circular)

def kadanes_all_negative_handling(arr):
    """
    Alternative implementation that handles all negative numbers explicitly
    
    Args:
        arr (list): Input array of integers
        
    Returns:
        int: Maximum sum of contiguous subarray
    """
    if not arr:
        return 0
    
    # If all numbers are negative, return the maximum single element
    if all(num < 0 for num in arr):
        return max(arr)
    
    max_so_far = 0
    max_ending_here = 0
    
    for num in arr:
        max_ending_here = max(0, max_ending_here + num)
        max_so_far = max(max_so_far, max_ending_here)
    
    return max_so_far

def print_test_case(arr, description):
    """Helper function to print test case results"""
    print(f"{description}: {arr}")
    result = kadanes_with_indices(arr)
    print(f"Maximum contiguous sum: {result['max_sum']}")
    print(f"Subarray: {result['subarray']} (indices {result['start_index']} to {result['end_index']})")
    print(f"Pythonic implementation: {kadanes_pythonic(arr)}")
    print()

def main():
    """Example usage and testing"""
    print("Kadane's Algorithm Examples:\n")
    
    # Test case 1: Mixed positive and negative numbers
    arr1 = [-2, -3, 4, -1, -2, 1, 5, -3]
    print_test_case(arr1, "Array 1 (mixed)")
    
    # Test case 2: All negative numbers
    arr2 = [-2, -3, -1, -5]
    print_test_case(arr2, "Array 2 (all negative)")
    
    # Test case 3: All positive numbers
    arr3 = [1, 2, 3, 4, 5]
    print_test_case(arr3, "Array 3 (all positive)")
    
    # Test case 4: Single element
    arr4 = [5]
    print_test_case(arr4, "Array 4 (single element)")
    
    # Test case 5: Circular array
    arr5 = [5, -3, 5]
    print(f"Array 5 (circular test): {arr5}")
    print(f"Maximum sum (linear): {kadanes_algorithm(arr5)}")
    print(f"Maximum sum (circular): {kadanes_circular(arr5)}")
    print()
    
    # Performance test with larger array
    import random
    large_arr = [random.randint(-10, 10) for _ in range(1000)]
    print(f"Large array test (1000 elements): Max sum = {kadanes_algorithm(large_arr)}")

if __name__ == "__main__":
    main()